sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.implementation.directconnectivity.rntbd;
+
+import com.azure.cosmos.implementation.AsyncDocumentClient;
+import com.azure.cosmos.implementation.Configs;
+import com.azure.cosmos.implementation.CosmosSchedulers;
+import com.azure.cosmos.implementation.IOpenConnectionsHandler;
+import com.azure.cosmos.implementation.OpenConnectionResponse;
+import com.azure.cosmos.implementation.ShouldRetryResult;
+import com.azure.cosmos.implementation.directconnectivity.Uri;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.ParallelFlux;
+import reactor.core.publisher.Sinks;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class ProactiveOpenConnectionsProcessor implements Closeable {
+
+    private static final Logger logger = LoggerFactory.getLogger(ProactiveOpenConnectionsProcessor.class);
+    private static final int MaxConnectionsToOpenAcrossAllEndpoints = 50_000;
+    private Sinks.Many<OpenConnectionOperation> openConnectionsTaskSink;
+    private Sinks.Many<OpenConnectionOperation> openConnectionsTaskSinkBackUp;
+    private final ConcurrentHashMap<String, Integer> endpointToMinConnections;
+    private final AtomicInteger totalEstablishedConnections;
+    private final AtomicReference<AsyncDocumentClient.OpenConnectionAggressivenessHint> aggressivenessHint;
+    private static final Map<AsyncDocumentClient.OpenConnectionAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+
+    static {
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(1, 1));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getCPUCnt(), Configs.getCPUCnt()));
+    }
+
+    public ProactiveOpenConnectionsProcessor() {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
```

Comment:
should we always flip to DEFENSIVE mode after the warmup or client is initialized?  Currently it will only be flipped if the customer call openConnectionAndInitCaches, what if they never do the warmup? 

(and to be honest, 1 probably is too low, it would be good to make these as configurable as well)

Reply:
This is a good one. I kept AGGRESSIVE as the default since the volume of open connection tasks when connections are reset / closed could be lower. But I could be wrong here.